camelus_lib.ml

```diff
@@ -931,6 +930,109 @@ module PrChecks = struct
       ]
     )
 
+  let upstream_check ancestor head gitstore =
+    let%lwt files = RepoGit.changed_files ancestor head gitstore in
+    let%lwt upstream_check =
+      Lwt_list.map_s (fun (file, contents) ->
+          let filename = OpamFilename.of_string file in
+          let opam =
+            OpamFile.OPAM.read_from_string ~filename:(OpamFile.make filename)
+              contents in
+          let upstream = OpamFile.OPAM.url opam in
+          let result =
+            match upstream with
+            | Some url ->
+              let open OpamProcess.Job.Op in
+              OpamProcess.Job.run @@
+              OpamFilename.with_tmp_dir_job @@ fun dir ->
+              OpamProcess.Job.catch (function
+                    Failure msg -> Done (`err msg)
+                  | e -> Done (`err (Printexc.to_string e)))
+              @@ fun () ->
+              OpamDownload.download ~overwrite:false (OpamFile.URL.url url) dir
+              @@| fun f ->
+              (match OpamFile.URL.checksum url with
+               | [] -> `warn
+               | chks ->
+                 let not_corresponding =
+                   List.filter (fun chk ->
+                       not( OpamHash.check_file (OpamFilename.to_string f) chk))
+                     chks
+                 in
+                 if not_corresponding = [] then
+                   `succ
+                 else
+                   let msg =
+                     Printf.sprintf "Cheksum%s %s don't verify archive"
+                       (if List.length chks = 1 then "" else "s")
+                       (OpamStd.List.to_string OpamHash.to_string not_corresponding)
+                   in
+                   `err msg
+              )
+            | None -> `fatal
+          in
+          let nv =
+            match OpamPackage.of_filename filename with
+            | Some nv -> nv
+            | None -> OpamPackage.of_string "invalid-package-name.v"
+          in
+          Lwt.return (nv, result)
+        ) (opam_files files)
+    in
+    let ok =
+      OpamStd.List.filter_map (fun (nv,res) ->
+          match res with
+          | `succ -> Some nv
+          | _ -> None) upstream_check
+    in
+    let warn =
+      OpamStd.List.filter_map (fun (nv,res) ->
+          match res with
+          | `warn -> Some nv
+          | _ -> None) upstream_check
+    in
+    let err =
+      OpamStd.List.filter_map (fun (nv,res) ->
+          match res with
+            `err m -> Some (nv, m)
+          | _ -> None) upstream_check
+    in
+    let fatal =
+      OpamStd.List.filter_map (fun (nv,res) ->
+          match res with
+          | `fatal -> Some nv
+          | _ -> None) upstream_check
+    in
+    let title, status =
+      if err <> [] || fatal <> []then
```

Comment:
why not display both upstream and lint warnings if we have both ? I would skip the `else`s here.